src/allocators/caller-guarantees.md

```diff
@@ -0,0 +1,69 @@
+## Guarantees to the allocator consumer
+
+This section describes properties that consumers of memory allocators may
+rely on from all CHERI-enabled allocators.
+
+### Allocating memory
+
+Calls to `malloc()` and `calloc() must return capabilities that:
+
+ * Are valid (i.e., with its tag bit set)
+ * Are unsealed
+ * Have bounds that permit access to the full requested memory range of the
+   allocation
+ * Have bounds that do not permit access to any other current allocations,
+   implementing non-aliasing spatial safety
+ * Have permissions that allow data load, data store, capability load, and
+   capability store
+ * Are sufficiently aligned to allow capability loads and stores at relative
+   offset 0 from the returned pointer
+
+The allocator may:
+
+ * Fill reachable memory within bounds with zeroes before returning a pointer
+   to it
+ * Provide precise bounds, with the lower bound being the bottom address of
+   the allocation, and the upper bound being one byte above the top address of
+   the allocation
+
+### Freeing memory
+
+The caller must not pass as an argument to `free()` a capability that:
+
+ * Is invalid (i.e., without its tag bit set)
+ * Is unsealed
+ * Has bounds other than those on the original capability returned by
+   `malloc()`, `calloc()`, or `realloc()`
+ * Has permissions that differ from those on the original capability returned
+   by `malloc()`, `calloc()`, or `realloc()`
+
+The allocator must not:
+
+ * Reuse storage associated with the allocation until there are no outstanding
+   valid capabilities that authorize access to the memory
+
+The allocator may:
+
+ * Fill reachable memory within the bounds of the allocation with zeroes after
+   it has been freed
```

Comment:
FWIW, if revocation is not prompt (CHERIoT, MTE), these zeros are advisory (and the allocator should re-zero memory before returning it from malloc).